### PR TITLE
planner: don't extract hash keys from index join's OtherConds if inl_merge_join hint exists

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/24954

Problem Summary:

`inl_merge_join` hint does not take effect.

### What is changed and how it works?

What's Changed:

Don't extract hash keys from index join's OtherConds if inl_merge_join hint exists.

How it Works:

In order to solve https://github.com/pingcap/tidb/issues/20710, https://github.com/pingcap/tidb/pull/20761 firstly extracts hash keys from `OtherConditions` when building `IndexJoin`, while https://github.com/pingcap/tidb/pull/21138 further prevents planner from generating index merge joins if hash keys have been extracted from `OtherConditions` since index merge join does not has the hash step at all. Hence no index merge join plan can be generated even if hint is specified sometimes.

This PR would check if `inl_merge_join` hint is specified before extracting the hash keys from `OtherConditions`, hence keep the chance to convert the result of `constructIndexJoin` to index merge join. 

### Related changes

N/A

### Check List 

Tests 

- Unit test

Side effects

N/A

### Release note 

- Don't extract hash keys from index join's OtherConds if inl_merge_join hint exists

### CODE REVIEW

**Comment by hi-rustin** [comment]
/lgtm

(Test the robot)

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -400,6 +400,7 @@ func (p *LogicalJoin) constructIndexJoin(
 	keyOff2IdxOff []int,
 	path *util.AccessPath,
 	compareFilters *ColWithCmpFuncManager,
+	extractOtherEQ bool,
```
Comment (winoros) [inline planner/core/exhaust_physical_plans.go:403]: Maybe add some comments would help others understand

**Review by winoros — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**